vidarr-niassa/src/main/java/ca/on/oicr/gsi/vidarr/niassa/NiassaOutputProvisioner.java

```diff
@@ -0,0 +1,130 @@
+    private final int[] chunks;
+    private final SSHClient client;
+    private final SFTPClient sftp;
+
+    private final Map<String,String> SUBSTITUTIONS = Map.of();
+
+    static final ObjectMapper MAPPER = new ObjectMapper();
+
+
+
+    // all these are for the TARGET. Specifically 'chunks' is the target. SOURCE comes from workflow? I think?
+    public NiassaOutputProvisioner(int[] chunks, String username, String hostname, short port) throws IOException {
+        this.chunks = chunks;
+        client = new SSHClient();
+        client.loadKnownHosts();
+        client.connect(hostname, port);
+        client.authPublickey(username);
+        sftp = client.newSFTPClient();
+    }
+
+    @Override
+    public boolean canProvision(OutputProvisionFormat format) {
+        return format == OutputProvisionFormat.FILES;
+    }
+
+    @Override
+    public void configuration(SectionRenderer sectionRenderer) throws XMLStreamException {
+        //skip
+    }
+
+    @Override
+    public JsonNode preflightCheck(JsonNode metadata, WorkMonitor<Boolean, JsonNode> monitor) {
+        // copy from CromwellOutputProvisioner
+        monitor.scheduleTask(() -> monitor.complete(true));
+        return null;
+    }
+
+
+    @Override
+    public void preflightRecover(JsonNode state, WorkMonitor<Boolean, JsonNode> monitor) {
+        // copy from CromwellOutputProvisioner
+        monitor.scheduleTask(() -> monitor.complete(true));
+    }
+
+    @Override
+    public JsonNode provision(String workflowRunId,
+                              String data, // this is a json object inside a string, containing labels md5 &c. this comes from workflowengine
+                              JsonNode metadata, // jsonobject, contains root of TARGET. Append chunks to this. this is the 'outputDirectory' defined in typeFor and comes from shesmu
+                              WorkMonitor<Result, JsonNode> monitor) {
+        monitor.scheduleTask(() -> {
+            // Set up all the data in the formats we need
+            JsonNode dataAsJson;
+            try {
+                dataAsJson = MAPPER.readTree(data);
+            } catch (JsonProcessingException e) {
+                throw new RuntimeException(e);
+            }
+            String sourcePath = dataAsJson.get("path").asText();
+            Path targetPath = Path.of(metadata.get("outputDirectory").asText());
+            int startIndex = 0;
+            for (final int length: chunks){
+                if (length < 1) break;
+                final int endIndex = Math.min(workflowRunId.length(), startIndex + length);
+                if (endIndex == startIndex) break;
+                targetPath = targetPath.resolve(workflowRunId.substring(startIndex, endIndex));
+                startIndex = endIndex;
+            }
+
+            //use the ssh connection to create symlink to the TARGET
+            try {
+                sftp.symlink(sourcePath, targetPath.toString());
```

Comment:
You have to create the parent directories first. I believe `sftp.mkdirs` can be used to recursively create the directory.

Reply:
Done!